x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/InferenceResolver.java

```diff
@@ -0,0 +1,213 @@
+            InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+                TaskType.ANY,
+                semanticQuery.inferenceId(),
+                null,
+                List.of(semanticQuery.queryString()),
+                Map.of(),
+                InputType.SEARCH,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API,
+                false
+            );
+
+            executeAsyncWithOrigin(
+                client,
+                ML_ORIGIN,
+                InferenceAction.INSTANCE,
+                inferenceRequest,
+                actionListener.delegateFailureAndWrap((next, inferenceResponse) -> {
+                    InferenceResults inferenceResults = SemanticTextInferenceUtils.validateAndConvertInferenceResults(
+                        inferenceResponse.getResults(),
+                        semanticQuery.fieldName()
+                    );
+                    inferenceResultsMap.put(semanticQuery, inferenceResults);
+                    next.onResponse(inferenceResponse);
+                })
+            );
+        }
+    }
+
+    private record SemanticQuery(String fieldName, String queryString, String inferenceId) {};
+
+    private Set<SemanticQuery> semanticQueries(LogicalPlan analyzedPlan, Set<Failure> failures, boolean isCrossClusterSearch) {
+        Set<SemanticQuery> result = new HashSet<>();
+
+        Set<String> indexNames = indexNames(analyzedPlan);
+
+        analyzedPlan.forEachExpressionDown(Match.class, match -> {
+            if (match.field().dataType() == DataType.SEMANTIC_TEXT && match.field() instanceof FieldAttribute field) {
+                if (isCrossClusterSearch) {
+                    failures.add(
+                        Failure.fail(
+                            match.field(),
+                            "[{}] {} does not allow semantic_text fields with cross cluster queries.",
+                            match.functionName(),
+                            match.functionType()
+                        )
+                    );
+                    return;
+                }
+
+                EsField esField = field.field();
+                Set<String> inferenceIds = inferenceIdsForField(esField.getName(), indexNames);
+                if (inferenceIds.size() == 1) {
+                    result.add(new SemanticQuery(field.sourceText(), match.query().sourceText(), inferenceIds.iterator().next()));
+                } else if (inferenceIds.size() == 0) {
+                    failures.add(
+                        Failure.fail(
+                            match.field(),
+                            "[{}] {} needs a configured inference ID but none could be found for [{}]",
+                            match.functionName(),
+                            match.functionType(),
+                            match.field().sourceText()
+                        )
+                    );
+                } else {
+                    failures.add(
```

Comment:
After thinking about it more, I wonder if we should allow multiple inference ids on the same field, if the inference ids come from different indices, and each index has only one inference id on this field. This is more like a usability issue.

If I have multiple indices that have the same semantic text field name, I couldn't find a way to query them in at once. Does it look like a valid use case? Schema is [here](https://github.com/elastic/elasticsearch/pull/116253#issuecomment-2512667746).
```
Thanks for fixing this!
+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from testidx00, testidx01 | where match(textfield, \"live\")"
}
'

{
  "error" : {
    "root_cause" : [
      {
        "type" : "verification_exception",
        "reason" : "Found 1 problem\nline 1:41: [MATCH] function cannot operate on [textfield] because it is configured with multiple inference IDs."
      }
    ],
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:41: [MATCH] function cannot operate on [textfield] because it is configured with multiple inference IDs."
  },
  "status" : 400
}

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from testidx00, testidx01 | where match(textfield::keyword, \"live\") | eval k = textfield::keyword | keep k"
}
'

{
  "error" : {
    "root_cause" : [
      {
        "type" : "verification_exception",
        "reason" : "Found 1 problem\nline 1:41: [MATCH] function cannot operate on [textfield::keyword], which is not a field from an index mapping"
      }
    ],
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:41: [MATCH] function cannot operate on [textfield::keyword], which is not a field from an index mapping"
  },
  "status" : 400
}
```

Reply:
@fang-xing-esql , the problem with multiple inference IDs is relevance. It does not make sense to mix multiple inference IDs from the scoring point of view, as different models will get different scores that can't directly be combined. That's the main reason it is not allowed on the `_search` endpoint.

In _search, users can combine different inference ids using different query clauses in _search, that can the be further adjusted via boosting.

Even if it's technically possible to do this from a filtering PoV, it won't make sense to do it once we combine with score.

Reply:
Can we take this as a follow up item?
Since this restriction is also present in _search, for all the reasons @carlosdelest mentioned, I thought having it in ES|QL is reasonable for now.
The time where it becomes relevant to lift this restriction is when we introduce RRF, in that case we might want to allow RRF between two subqueries that query the same field on different indices: `WHERE where match(textfield, "live") AND _index == "testidx00"` and `WHERE where match(textfield, "live") AND _index == "testidx01"`.

Reply:
Yes, this can be addressed as a follow up.